File: src/java/magmac/compile/PrefixRule.java
package magmac.compile;

import java.util.Optional;

public record PrefixRule(
        String prefix, Rule rule
) implements Rule {
    @Override
    public Optional<MapNode> parse(String input) {
        if (input.startsWith(this.prefix())) {
            var slice = input.substring(this.prefix().length());
            return this.rule().parse(slice);
        }

        return Optional.empty();
    }

    @Override
    public Optional<String> generate(MapNode node) {
        return this.rule.generate(node).map(inner -> this.prefix + inner);
    }
}

File: src/main/windows/magma/compile/rule/PrefixRule.c
struct PrefixRule(String prefix, Rule rule) implements Rule{};
Rule Prefix_PrefixRule(String prefix, Rule rule) implements Rule(String prefix,  Rule rule) {}
CompileError> lex_PrefixRule(String prefix, Rule rule) implements Rule(String content) {}
CompileError> generate_PrefixRule(String prefix, Rule rule) implements Rule(Node node) {}
